elyra/pipeline/component_catalog.py

```diff
@@ -25,25 +29,133 @@
 from elyra.metadata.schemaspaces import ComponentCatalogs
+from elyra.pipeline.catalog_connector import ComponentCatalogConnector
 from elyra.pipeline.component import Component
 from elyra.pipeline.component import ComponentParser
+from elyra.pipeline.component_metadata import ComponentCatalogMetadata
+from elyra.pipeline.runtime_type import RuntimeProcessorType
 
+BLOCKING_TIMEOUT = 5
+NONBLOCKING_TIMEOUT = 0.10
+CATALOG_UPDATE_TIMEOUT = 15
 
-class ComponentCatalog(LoggingConfigurable):
-    """
-    Component Catalog, responsible to provide a list of available components
-    for each runtime. The catalog uses component parser to read and parse each
-    component entry from the catalog and transform them into a component value object.
-    """
-    _cached_components: Dict[str, Component] = {}
-    _cache_last_updated = None
+
+class CacheUpdateManagerThread(Thread):
+    def __init__(self, log: Logger, component_cache: Dict[str, Dict], cache_task_queue: Queue):
+        super().__init__()
+
+        self.setDaemon(True)
+
+        self.log = log
+        self._component_cache = component_cache
+        self._cache_task_queue = cache_task_queue
+
+        self._threads: List[CacheUpdateThread] = []
+
+        self.stop_event = Event()
+
+    def run(self):
+        outstanding_threads = False
+
+        while not self.stop_event.is_set():
+            timeout = NONBLOCKING_TIMEOUT if outstanding_threads else BLOCKING_TIMEOUT
+            try:
+                # Get a task from the queue
+                catalog, operation = self._cache_task_queue.get(timeout=timeout)
+
+            except Empty:
+                # No task exists in the queue, proceed to check for thread execution
+                pass
+
+            else:
+                # Create and start a thread for the task
+                updater_thread = CacheUpdateThread(self._component_cache, self._cache_task_queue, catalog, operation)
+                updater_thread.start()
+
+                self._threads.append(updater_thread)
+
+            outstanding_threads = self.check_outstanding_threads()
+
+    def check_outstanding_threads(self) -> bool:
+        """
+        Join finished threads and report on long-running threads as needed.
+        """
+        outstanding_threads = False
+        for thread in self._threads:
+            # Attempt to join thread within the given amount of time
+            thread.join(timeout=NONBLOCKING_TIMEOUT)
+
+            if thread.is_alive():
+                # Thread is still running (thread join timed out)
+                outstanding_threads = True
+                duration = int(time.time() - thread.last_warn_time)
+                if duration > CATALOG_UPDATE_TIMEOUT:
+                    thread.last_warn_time = time.time()
+                    self.log.warning(f"Cache update for catalog '{thread.name}' is still "
+                                     f"processing after {duration} seconds ...")
```

Comment:
I think the duration here should be cumulative...
```suggestion
                                     f"processing after {int(thread.last_warn_time - thread.task_start_time)} seconds ...")
```

Reply:
Good catch!